packages/contracts/contracts/votings/majority-voting/WhitelistVoting.sol

```diff
@@ -0,0 +1,175 @@
+        _addWhitelistedUsers(_whitelisted);
+    }
+
+    /// @notice add new users to the whitelist.
+    /// @param _users addresses of users to add
+    function addWhitelistedUsers(address[] calldata _users) external auth(MODIFY_WHITELIST) {
+        _addWhitelistedUsers(_users);
+    }
+
+    /// @dev Internal function to add new users to the whitelist.
+    /// @param _users addresses of users to add
+    function _addWhitelistedUsers(address[] calldata _users) internal {
+        for (uint256 i = 0; i < _users.length; i++) {
+            whitelisted[_users[i]] = true;
+        }
+
+        whitelistedLength += uint64(_users.length);
+
+        emit AddUsers(_users);
+    }
+
+    /// @notice remove new users to the whitelist.
+    /// @param _users addresses of users to remove
+    function removeWhitelistedUsers(address[] calldata _users) external auth(MODIFY_WHITELIST) {
+        for (uint256 i = 0; i < _users.length; i++) {
+            whitelisted[_users[i]] = false;
+        }
+
+        whitelistedLength -= uint64(_users.length);
+
+        emit RemoveUsers(_users);
+    }
+
+    /// @notice Create a new vote on this concrete implementation
+    /// @param _proposalMetadata The IPFS hash pointing to the proposal metadata
+    /// @param _actions the actions that will be executed after vote passes
+    /// @param _startDate state date of the vote. If 0, uses current timestamp
+    /// @param _endDate end date of the vote. If 0, uses _start + minDuration
+    /// @param _executeIfDecided Configuration to enable automatic execution on the last required vote
+    /// @param _choice Vote choice to cast on creationr
+    function newVote(
+        bytes calldata _proposalMetadata,
+        IDAO.Action[] calldata _actions,
+        uint64 _startDate,
+        uint64 _endDate,
+        bool _executeIfDecided,
+        VoterState _choice
+    ) external override returns (uint256 voteId) {
+        if(!whitelisted[_msgSender()]) revert VoteCreationForbidden(_msgSender());
+
+        // calculate start and end time for the vote
+        uint64 currentTimestamp = getTimestamp64();
+
+        if (_startDate == 0) _startDate = currentTimestamp;
+        if (_endDate == 0) _endDate = _startDate + minDuration;
+
+        if(_endDate - _startDate <  minDuration || _startDate < currentTimestamp)
+            revert VoteTimesForbidden({
+                current: currentTimestamp,
+                start: _startDate,
+                end: _endDate,
+                minDuration: minDuration
+            });
+
+        voteId = votesLength++;
+
+        // create a vote.
+        Vote storage vote_ = votes[voteId];
+        vote_.startDate = _startDate;
+        vote_.endDate = _endDate;
+        vote_.supportRequiredPct = supportRequiredPct;
+        vote_.participationRequiredPct = participationRequiredPct;
+        vote_.votingPower = whitelistedLength;
+
+        for (uint256 i; i < _actions.length; i++) {
+            vote_.actions.push(_actions[i]);
```

Comment:
don't want to annoy but if we anyways touch this could we directly put it into the ``unchecked`` closure and store the actions length on the stack to save some gas.